pkg/liqo-controller-manager/ip-controller/ip_controller.go

```diff
@@ -0,0 +1,203 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ipctrl
+
+import (
+	"context"
+	"fmt"
+	"net"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	ipamv1alpha1 "github.com/liqotech/liqo/apis/ipam/v1alpha1"
+	"github.com/liqotech/liqo/pkg/consts"
+	"github.com/liqotech/liqo/pkg/liqonet/ipam"
+	foreignclusterutils "github.com/liqotech/liqo/pkg/utils/foreignCluster"
+)
+
+const (
+	ipamIPFinalizer = "ip.ipam.liqo.io"
+)
+
+// IPReconciler reconciles a IP object.
+type IPReconciler struct {
+	client.Client
+	Scheme     *runtime.Scheme
+	IpamClient ipam.IpamClient
+}
+
+// +kubebuilder:rbac:groups=ipam.liqo.io,resources=ips,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=ipam.liqo.io,resources=ips/status,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=ipam.liqo.io,resources=ips/finalizers,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=ipam.liqo.io,resources=networks,verbs=get;list;watch
+// +kubebuilder:rbac:groups=discovery.liqo.io,resources=foreignclusters,verbs=get;list;watch
+
+// Reconcile Ip objects.
+func (r *IPReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var ip ipamv1alpha1.IP
+	var desiredIP, remappedIP string
+
+	// Fetch the IP instance
+	if err := r.Get(ctx, req.NamespacedName, &ip); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.V(4).Infof(" %q not found", req.NamespacedName)
+			return ctrl.Result{}, nil
+		}
+		klog.Errorf("an error occurred while getting IP %q: %v", req.NamespacedName, err)
+		return ctrl.Result{}, err
+	}
+
+	// Retrieve the remote cluster ID from the labels.
+	remoteclusterID := ip.Labels[consts.RemoteClusterID] // it should always be present thanks to validating webhook
+
+	_, err := foreignclusterutils.GetForeignClusterByID(ctx, r.Client, remoteclusterID)
+	if apierrors.IsNotFound(err) {
+		klog.Errorf("no ForeignCluster with clusterID %q found: %v", remoteclusterID, err)
+		return ctrl.Result{}, err
+	} else if err != nil {
+		klog.Errorf("an error occurred when trying to retrieve ForeignClusters: %v", err)
+		return ctrl.Result{}, err
+	}
+
+	desiredIP = ip.Spec.IP
+
+	if ip.GetDeletionTimestamp().IsZero() {
+		if !controllerutil.ContainsFinalizer(&ip, ipamIPFinalizer) {
```

Comment:
Can you move the if/else content in functions to avoid high level of indentation that makes it diffucult to read?